Sources/Datadog/RUM/RUMVitals/VitalRefreshRateReader.swift

```diff
@@ -54,7 +60,12 @@ internal class VitalRefreshRateReader {
         if let lastTimestamp = self.lastFrameTimestamp {
             let frameDuration = link.timestamp - lastTimestamp
             let currentFPS = 1.0 / frameDuration
-            valuePublisher.publishAsync(currentFPS)
+
+            for publisher in valuePublishers {
+                publisher.mutateAsync { currentInfo in
+                    currentInfo.addSample(currentFPS)
+                }
+            }
```

Comment:
Why not having one single publisher for the `currentFPS` value? This way we won't need to run blocks on `DispatchQueue.main` for every subscription / unsubscription (for every new view started) and the overall threading model could be simpler:

```swift
// In a single instance of refresh rate reader; calling this non-blocking "publish" on the main thread:
publisher.publishAsync(currentFPS)

// In each sampler (one sampler per view); every 1 second; calling this blocking read from `RUMViewScope` thread:
let lastFPS = publisher.value
```

Reply:
For what I see - this subscribe / unsubscribe introduces [the leak](https://github.com/DataDog/dd-sdk-ios/pull/530#discussion_r666059664) which makes it worth having a simpler threading model around the `FPS` metric.

Reply:
`VitalInfoSampler` also needs to `refreshRateReader.unregister(refreshRatePublisher)` in `deinit()`, this was another bug. fixing this now 👍

Reply:
memory leak comes from the `timer`. [Apple docs](https://developer.apple.com/documentation/foundation/timer/1408356-init) says
> `target`: The timer maintains a strong reference to this object until it (the timer) is invalidated.

`VitalInfoSampler` retains `timer` as instance variable and `timer` retains `VitalInfoSampler` as `target`. i'm looking into it 👀

Reply:
here `ValuePublisher` doesn't actually _publish_ anything reactively.
it is used as a wrapper for one single `value: VitalInfo` so that every `RUMViewScope` will create its own subscriber and will be getting its own value.
if there are 5 different `RUMViewScope`s at once, each needs different `VitalInfo` values; so each needs to have its own `ValuePublisher`.

if we want to use a single publisher it will be `reader -> reader's publisher -> publisher's publishers -> receiver`.
i can't think of a case where `DispatchQueue.main.async { array.append(element) }` can cause any issues, do you have a case in your mind?

Reply:
✅ retain cycle is fixed here: https://github.com/DataDog/dd-sdk-ios/pull/530/files#diff-f15ba741c3c811beb0029e017dd16bd5271a7d25c1d0d819a2f0de509fb271d9R61-R66

```swift
let timer = Timer(...) { [weak self] _ in
  self?.takeSample()
}
```

Reply:
>  can't think of a case where `DispatchQueue.main.async { array.append(element) }` can cause any issues, do you have a case in your mind?

Its current impact is negligible (one sync for each view transition), but this model would need to be changed if we want to add more things there in the future. It's OK for now if we consider this not harmful 👍.